[ICDS Dashboard]Added the Migration Form Model

ShubhamBansal1997 opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Their PR accounts for the Migration Form Model


##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

ShubhamBansal1997 committed:
[build]: Migration Form Model
- const.py -> added the constants for the table
- aggreate.py -> model for migration table
- migration_form.py -> migration helper

- JIRA Ticket:
    https://dimagi-dev.atlassian.net/browse/ICDS-1025
- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

Rohit25negi commented on custom/icds_reports/models/aggregate.py:
```diff
@@ -1520,3 +1522,33 @@ class Meta(object):
+    month = models.DateField(help_text="Will always be YYYY-MM-01")
+
+    # not the real pkey - see unique_together
+    case_id = models.CharField(max_length=40, primary_key=True)
```
Will be clearer to say it as person_case_id as its coming from migration form and person case id is pulled from there so that its not confused with any other case id

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -0,0 +1,49 @@
+from dateutil.relativedelta import relativedelta
```
Can you fix lynting errors?

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -0,0 +1,49 @@
+        return """
+        INSERT INTO "{tablename}" (
+        	case_id, state_id, supervisor_id, month, latest_time_end_processed,
+			migration_status
```
Where are you fetching migration_Status from? it looks its missed to be fetched in the select sub query?

Rohit25negi reviewed (changes requested):
Left a few comments.
Also can you change the base of this branch to the UCR branch you created?

ShubhamBansal1997 committed:
[fix]: minor linting fixes

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

mkangia commented:
Build failure looks related @ShubhamBansal1997

ShubhamBansal1997 committed:
[fix]: include the MigrationFormsAggregationDistributedHelper in the
__init__.py file and correct the unique_together in the django model definition

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[migration]: added the migratio file

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: remove extra autogenerated migration code

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

Rohit25negi reviewed (changes requested):
Left a few comments.

ShubhamBansal1997 committed:
[fix]: minor fixes
- removed extra field from the migration model last_time_processed
- taking previous month data into account while getting data from the
UCR

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: removed extra field

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[refactor]: updated the migration_status in the CCSRecordMonthlyHelper
from the person_cases to migration_table

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[build]: Updated the aggregation output due to addition of the
miigration form

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: resovle the migration merge issues

django.core.management.base.CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0158_auto_20191227_1147, 0159_update_agg_awc_monthly_view_visit_fields in icds_reports).

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[refactor]: change the type of the migration_status to postive int in
-> ucr
-> model
-> migration

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: made the AggregateMigrationForms distributed

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: fix a failing test custom.icds_reports.tests.agg_tests.reports.test_demographics:TestDemographics.test_data_daily_if_aggregation_script_fail_lactating_enrolled_women

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

calellowitz commented:
seeing lots of test failures

calellowitz commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -0,0 +1,81 @@
+          WHERE state_id = %(state_id)s AND
+                timeend >= %(current_month_start)s AND timeend < %(next_month_start)s AND
+                person_case_id IS NOT NULL
+          WINDOW w AS (
```
you dont appear to use this window at all. what is it for?

calellowitz reviewed (changes requested):
Few concerns and questions
Also, this also needs to be used for child health cases, right?

Rohit25negi commented on custom/icds_reports/ucr/data_sources/migrations_form.json:
```diff
@@ -143,21 +143,25 @@
       {
         "column_id": "migration_status",
-        "datatype": "string",
+        "datatype": "small_integer",
```
I believe this would be better to leave like string for the case as handling this migration would be tedious. The value could be just either `migrated` or null. So aggregation just check for value `migrated`.  @calellowitz ?

calellowitz replied:
Depending on how many migration forms there are, the ideal might be to make the change then drop everything and reindex from scratch. If there are too many forms we can make a new column that holds this as an integer, and leave the existing column but not use it. Either way, an integer is preferable because it keeps this table consistent with other UCRs

Rohit25negi replied:
My preference would to be then to create a new column. COol. THanks,

Rohit25negi replied:
@ShubhamBansal1997  could you create a new column in the UCR which will have value as Integer and use that further in aggregation instead of this existing one.  you can name it as is_migrated something
I think we can also remove this existing column(please wait, I am confirming something). 

But you can create a new one holding ints and use that in agg.

ShubhamBansal1997 committed:
[fix]: minor name fix agg_min -> agg_migration
migration_status default to 0 instead of null

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[revert]: reverting back the adding of the migration model in the ccs
record monthy will do after merging of this PR

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[revert]: since the UCR is migration, so values are already stored as
string, which are difficult to migrate

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[refactor]: shifting migrations_form.json to dashboard since it's being
used in the dashboard

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[refactor]: converting migration_status into 1, 0 forms at the agg level

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[revert]: reverting test changes

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[revert]: revert back the ccs record sql

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: fix the test_migration_form

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

Rohit25negi commented on custom/icds_reports/ucr/data_sources/migrations_form.json:
```diff
@@ -143,21 +143,25 @@
       {
         "column_id": "migration_status",
-        "datatype": "string",
+        "datatype": "small_integer",
```
I have already asked Shubham to just create a new column.
Just out of curiosity, will the migration still break if we would have changed the column name? Because I think in that db will delete the existing column and create a new one? 
Also could that cause any problem like locking etc?

Rohit25negi replied:
@calellowitz

ShubhamBansal1997 committed:
[fix]: added new column is_migrated and using the same for handling the
migration_status

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

Rohit25negi commented on custom/icds_reports/tests/agg_tests/outputs/agg_awc_2017-04-01_2_sorted.csv:
```diff
@@ -3,4 +3,4 @@ st1,d1,All,All,All,2017-04-01,26,130,,,,,,,,,,,,,,,,,12,,338,497,,135,0,,,,,,,,5
 st2,d3,All,All,All,2017-04-01,12,30,,,,,,,,,,,,,,,,,4,,196,270,,58,0,,,,,,,,26,33,351,1,0,0,0,0,0,0,0,0,0,0,0,0,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,2,1,2,,0,1,2,2,35,1,0,yes,,,,,,2,1,1,1,2,5,610,257,259,,26,33,351,9,3,9,3,2,,,,2,73,410,0,0,0,2,0,0,0,0,0,0,0,1,0,111,39,33,,59,0,,,2,2,2,6,0,33,26,33,26
 st6,d7,All,All,All,2017-04-01,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,0,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,yes,,,,,,2,1,1,1,1,1,1,,,,0,1,,,,,,,,,,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,1,,,,1,0,,,,,0,,,1,0,1,0
-st7,d8,All,All,All,2017-04-01,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,0,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,yes,,,,,,2,1,1,1,1,1,1,0,4,,0,1,,0,0,4,0,,,,,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,1,,,,1,0,,,,,0,3,2,1,0,1,0
+st7,d8,All,All,All,2017-04-01,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,0,1,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,yes,,,,,,2,1,1,1,1,1,1,0,4,,0,1,,0,0,4,0,,,,,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,1,,,,1,0,,,,,0,3,2,1,0,1,0
```
good not to change the files when there is not actual change.

Rohit25negi commented on custom/icds_reports/models/aggregate.py:
```diff
@@ -1549,4 +1551,30 @@ class Meta(object):
+    # not the real pkey - see unique_together
+    person_case_id = models.CharField(max_length=40, primary_key=True)
+
+    migration_status = models.PositiveSmallIntegerField(blank=True, null=True, help_text="Status of the Migration")
```
good to make its name same as UCr.

Rohit25negi reviewed (approved):
Looks good from my end. Left a few good practice comments But not blocking at all.
Will leave to Cal if he has final questions.

ShubhamBansal1997 committed:
[fix]: minor fixes
- changing the name of the field migration_status to is_migrated

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: minor column name fix

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 commented:
@calellowitz Please review

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -0,0 +1,81 @@
+)
+
+
+class MigrationFormsAggregationDistributedHelper(StateBasedAggregationDistributedHelper):
```
you will also need to add this migration to the Airlfow Pipes [repo](https://github.com/dimagi/pipes).
You can do that by creating an aggregation function of this in [this file like other agg fuctions](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/tasks.py#L667) and first call that function in `move_ucr_data_into_aggregation_tables` function in similar way others are being called so that its being called during local test cases and then also add that function to [here](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/management/commands/run_aggregation_query.py) and also update the Pipes report. Example PR could be [this](https://github.com/dimagi/pipes/pull/24/files)

Rohit25negi reviewed (changes requested):
@ShubhamBansal1997 requested a few changes to make this complete. Sorry for not catching this earlier

Rohit25negi commented:
Also, Holding for merging. As there are some discussions going on around the inclusion and exclusion criteria of migrated cases.

ShubhamBansal1997 committed:
[chore]: adding the agg migration form tasks to airflow

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 commented:
@Rohit25negi updated the PR

Rohit25negi commented:
@ShubhamBansal1997  could you update the code to include the change migration calculation as mentioned in the ticket. Basically you could do it  by adding 'migration_date' column in your table and in subsequent tables you could check if the migration date is before the current month start

ShubhamBansal1997 committed:
[fix]: added the migration date for migration aggregation migration

- Github Issue:

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -29,7 +29,16 @@ def data_from_ucr_query(self):
+                    THEN LAST_VALUES(timeend)
+                ELSE LAST_VALUES(date_left) OVER w
+            END AS migration_date,
+            CASE
```
you don't have to have this column. Only plain one `LAST_VALUES(is_migrated) OVER w AS migration_status,` is fine.

This check we can do in following agg tables.

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -29,7 +29,16 @@ def data_from_ucr_query(self):
             %(month)s::DATE AS month,
             person_case_id AS person_case_id,
-            LAST_VALUES(is_migrated) OVER w AS is_migrated,
+            LAST_VALUES(is_migrated) OVER w AS migration_status,
```
It looks there are some columns name mixing up.

Rohit25negi commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -29,7 +29,16 @@ def data_from_ucr_query(self):
-            LAST_VALUES(is_migrated) OVER w AS is_migrated,
+            LAST_VALUES(is_migrated) OVER w AS migration_status,
+            CASE
+                WHEN LAST_VALUES(date_left) OVER w IS NULL
```
can you check the syntax. Looks  syntax error.

Rohit25negi reviewed (changes requested):
Left a few comment

ShubhamBansal1997 committed:
[fix]: minor fixes

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: minor test fixes and tasks fixes

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: minor typo fix

- Github Issue:

ShubhamBansal1997 committed:
[refactor]: removed too much migration file and transformed into one

- Github Issue:

ShubhamBansal1997 committed:
[fix]: remove the extra option

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: minor typo fix

- Github Issue:

ShubhamBansal1997 committed:
[revert]: changes in the model

- Github Issue:

ShubhamBansal1997 committed:
[fix]: minor fix in sql query

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

Rohit25negi commented on custom/icds_reports/ucr/data_sources/dashboard/migrations_form.json:
```diff
@@ -158,6 +158,46 @@
         "is_primary_key": false
+      },
+      {
+        "column_id": "is_migrated",
```
I am assuming these new fields are tested that they working fine and are able to fetch the required data from form?
@ShubhamBansal1997

Rohit25negi commented on custom/icds_reports/ucr/tests/test_migration_form.py:
```diff
@@ -14,8 +14,10 @@ def test_migration_form(self):
         self._test_data_source_results(
             'migration_form_v31895',
             [{
+                "date_left": None,
```
could you add a test form to check whether this new field added `date_left` is actually fetching the data from form.

Rohit25negi reviewed (commented):
Looks good to me. could you just add a test form for date_left field added?

ShubhamBansal1997 committed:
[refactor]: added more tests

- Github Issue:

Rohit25negi reviewed (approved):
Looks good. @ShubhamBansal1997  can you also add this agg step to pipes?

I will leave for @calellowitz  to give the final go-ahead. As he had left a few comments before on this.

ShubhamBansal1997 commented:
@Rohit25negi Added to pipes https://github.com/dimagi/pipes/pull/26

calellowitz reviewed (changes requested):
Few small things

ShubhamBansal1997 commented on custom/icds_reports/models/aggregate.py:
```diff
@@ -1551,6 +1554,33 @@ class Meta(object):
         unique_together = ('month', 'supervisor_id', 'person_case_id')  # pkey
 
     _agg_helper_cls = AggAdolescentGirlsRegistrationAggregate
+
```
I don't think so, might be done while taking hints from another example.

calellowitz replied:
Im not sure I understand this comment. Why did you remove it from that other class? If you did not intend to, you should add it back.

ShubhamBansal1997 replied:
Added back

ShubhamBansal1997 commented on custom/icds_reports/ucr/data_sources/dashboard/migrations_form.json:
```diff
@@ -158,6 +158,46 @@
         "is_primary_key": false
+      },
+      {
+        "column_id": "is_migrated",
```
Isn't better to remove the old column, since it's not used now?

calellowitz replied:
yeah, if you are going to rebuild the table anyway, removing the old column and replacing it with a new one is equivalent to changing the type. I guess its your call what you do, but consistency is naming columns is good so using whatever name we use elsewhere would be best.

ShubhamBansal1997 replied:
Removed `migration_status` column

ShubhamBansal1997 commented on custom/icds_reports/utils/aggregation_helpers/distributed/migration_forms.py:
```diff
@@ -0,0 +1,87 @@
+            %(month)s::DATE AS month,
+            person_case_id AS person_case_id,
+            LAST_VALUE(is_migrated) OVER w AS is_migrated,
+            CASE
```
`migration_date` is the main thing on which the migration status behaviour is working.It basically used to identify in which month actually a person is migrated instead of marking migrated in all months. https://dimagi-dev.atlassian.net/browse/ICDS-1025 Check Nitish comment on January 16, 2020, 4:37 AM

calellowitz replied:
Right, I mean the `CASE` statement, when would `date_left` be blank and we have to fall back to submission date?

ShubhamBansal1997 replied:
@Rohit25negi

Rohit25negi replied:
Yeah, date_left is a question in the form which is **not required**. It may sometime come as empty because aww does not date when the people left or with other reasons. That's why we have to fall back to timeend

ShubhamBansal1997 committed:
[refactor]: remove extra column

- Github Issue:

ShubhamBansal1997 committed:
[fix]: fix test as per the change in the migration ucr

- Github Issue:

ShubhamBansal1997 committed:
[revert]: add the param removed accidently

- Github Issue:

ShubhamBansal1997 committed:
Merge branch 'master' into sb/migration_model

ShubhamBansal1997 committed:
[fix]: migration fix

- Github Issue:

czue commented:
this was quite a journey. thanks for riding it through to the end! :motorcycle:
